packages/teleterm/src/ui/services/connectionTracker/connectionTrackerService.ts

```diff
@@ -129,29 +129,32 @@ export class ConnectionTrackerService extends ImmutableStore<ConnectionTrackerSt
       .disconnect();
   }
 
-  removeItem(id: string): void {
+  async removeItem(id: string): Promise<void> {
     const connection = this.state.connections.find(c => c.id === id);
     if (!connection) {
       return;
     }
 
-    this._trackedConnectionOperationsFactory.create(connection).remove?.();
+    await this._trackedConnectionOperationsFactory
+      .create(connection)
+      .remove?.();
 
     this.setState(draft => {
       draft.connections = draft.connections.filter(i => i.id !== id);
     });
-
-    this._statePersistenceService.saveConnectionTrackerState(this.state);
   }
 
-  removeItemsBelongingToRootCluster(clusterUri: string): void {
-    this.setState(draft => {
-      draft.connections = draft.connections.filter(i => {
+  async removeItemsBelongingToRootCluster(clusterUri: string): Promise<void> {
+    await Promise.all(
+      this.getConnections().map(async connection => {
         const { rootClusterUri } =
-          this._trackedConnectionOperationsFactory.create(i);
-        return rootClusterUri !== clusterUri;
-      });
-    });
+          this._trackedConnectionOperationsFactory.create(connection);
+        if (rootClusterUri === clusterUri) {
+          await this.disconnectItem(connection.id);
+          await this.removeItem(connection.id);
+        }
+      })
+    );
```

Comment:
I'm not fond of this change to be honest. Before this change, when the user was logging out, `connectionTrackerService` was responsible only for cleaning up its own state related to the cluster that the user is logging out from. It barely had any side effects, other than updating the state of the connection tracker.

Now `removeItemsBelongingToRootCluster` orchestrates the disconnection and removal of every active cluster item.

It also used to be a single update of connection tracker's state, while now every time an item is removed, it updates the workspace service which has repercussions on the whole app but it also runs `ConnectionTrackerService._refreshState` again on every removed item.

I realize it's convenient to use as `TrackedConnectionOperations` appears to include everything that we need to accomplish this goal. However, did you consider orchestrating that within `ClustersService`? It seems that on logout we need to remove cluster gateways  and kube configs. gravitational/webapps.e#311 is about removing gateways on tshd's side. Perhaps removing those kube configs could happen there as well?

Reply:
I agree with your points, calling `refreshState` on every removed item is not nice, but we can try to optimize it. 

However, I'm not sure how do you see orchestrating the resource removal in `ClustersService`. For example, `kubeConfigName` lives only within a connection, how the service will know about it?

Changing the way how `removeItemsBelongingToRootCluster` works seemed natural to me. When user clicks on disconnect/remove buttons we remove the resources, so why not do it when more than one connection is removed? Thanks to this we keep the cleanup logic in one place.

Reply:
> When user clicks on disconnect/remove buttons we remove the resources, so why not do it when more than one connection is removed?

Because it ultimately muddies up the logic behind logging out, I think. The act of removing a single connection is different from logging out. Before the change, during logout `ConnectionTrackerService` was responsible for merely updating its own state to reflect the changes from other services. Now it's responsible not only for its own state, but also the shutdown of whole resources (gateways and k8s).

I don't think such shift in responsibility is necessary. `ConnectionTrackerService` already does a lot, `_refreshState` is a complex function and I don't think we should put even more logic into the service. In my eyes, that service works well as a simple reflection of the "history" of added documents. Currently it knows that some resources exists (gateways, k8s) but ultimately it doesn't manage their lifetimes – this is done by the logic in `Document`-type components, in `ClustersService` and so on.

The `disconnect` function does more than what's necessary during logout. It closes and removes documents from the state which is something that's already being done by `ctx.workspacesService.removeWorkspace(clusterUri)`. It removes gateways which is something that we'd like to do on the backend when calling `ctx.clustersService.logout`.

> However, I'm not sure how do you see orchestrating the resource removal in `ClustersService`. For example, `kubeConfigName` lives only within a connection, how the service will know about it?

This could be handled by putting all kube configs belonging to a particular profile into a single directory or giving their filenames the same prefix so that during logout we know already which files we need to remove. If that's not possible due to time constraints, I think I'd still prefer `ConnectionTrackerService` to merely provide the names of the kube configs that should be removed and put the logic behind actually removing them somewhere else.

`useClusterLogout` already has access to the app context, so it can execute methods directly on the main process client. `ClustersService` seems to accept the main process client as a dependency only so that we don't have to pass the main process client as another dep to `TrackedConnectionOperationsFactory`, doesn't it?

Ultimately, `useClusterLogout` is the place that keeps the cleanup logic. I'd rather keep it explicitly expressed there. With `ConnectionTrackerService` calling `disconnect` on each item belonging to a root cluster, we bounce around between a couple of different services, cause a couple of updates because of the state changes. It was not the case before this change.

Reply:
Ok, I think I understand it finally, thanks for spending your time on writing things down!  
I decided to go with placing all kube configs belonging to a particular profile into a single directory. It looks like the simplest solution. 

>useClusterLogout already has access to the app context, so it can execute methods directly on the main process client. ClustersService seems to accept the main process client as a dependency only so that we don't have to pass the main process client as another dep to TrackedConnectionOperationsFactory, doesn't it?

Yes, so maybe the best place to put the clean up logic is `ClustersService` (as you proposed in the first comment)?
I expanded the `logout` method to remove the cluster and kube configs. 
However, I can move the cleanup logic to `useClusterLogout` if you are conceived that is better.

Do you have any opinion about the naming? I changed `kubeConfigName`param to `relativePath` in `removeKubeConfig()`, because now we can remove either `localhost/my-kube-abc` or `localhost` as a dir. 
For a kube document I think I will leave it as it is (`kubeConfigName`).